include/inexor/vulkan-renderer/io/byte_stream.hpp

```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include <cstdint>
+#include <filesystem>
+#include <vector>
+
+namespace inexor::vulkan_renderer::io {
+class ByteStream {
+protected:
+    std::vector<std::uint8_t> m_buffer;
+
+    /// Read from file.
+    [[nodiscard]] static std::vector<std::uint8_t> read_file(const std::filesystem::path &path);
+
+public:
+    ByteStream() = default;
+    explicit ByteStream(std::vector<std::uint8_t> buffer);
+    /// Read from file.
+    explicit ByteStream(const std::filesystem::path &path);
+
+    [[nodiscard]] std::size_t size() const;
+    [[nodiscard]] const std::vector<std::uint8_t> &buffer() const;
+};
+
+class ByteStreamReader {
+private:
+    const ByteStream &m_stream;
+    /// Stream iterator.
+    std::vector<std::uint8_t>::const_iterator m_iter;
+
+    void check_end(std::size_t size) const;
+
+public:
+    explicit ByteStreamReader(const ByteStream &stream);
+
+    [[nodiscard]] std::size_t remaining() const;
+    void skip(std::size_t size);
+
+    /// Generic read method.
+    template <class T, class... Args>
```

Comment:
Since the cli parser code uses `typename`, we should probably use it here too, or at least switch all of the other templates to use `class`, to be consistent. This is just my personal opinion, but I think `typename` looks nicer than `class` in templates as `T` can be any type not just a `class`.